### ISSUE TITLE
Support async/await

### ISSUE THREAD

**Issue by sindresorhus** [onevcat/Kingfisher#1864]
### Check List

Thanks for considering to open an issue. Before you submit your issue, please confirm these boxes are checked.

- [x] I have read the [wiki page](https://github.com/onevcat/Kingfisher/wiki) and [cheat sheet](https://github.com/onevcat/Kingfisher/wiki/Cheat-Sheet), but there is no information I need.
- [x] I have searched in [existing issues](https://github.com/onevcat/Kingfisher/issues?utf8=✓&q=is%3Aissue), but did not find a same one.
- [x] I want to report a problem instead of asking a question. It'd better to use [kingfisher tag in Stack Overflow](http://stackoverflow.com/questions/tagged/kingfisher) to ask a question.

### Issue Description

Swift will get [back-deployable async/await](https://developer.apple.com/documentation/xcode-release-notes/xcode-13_2-release-notes) with Xcode 13.2 (which is in beta). It would be great if Kingfisher would adopt async/await so it would be easier to use it from an async context.

If you're not ready to rework everything using async/await, at least consider providing some async overloads.

Example:

```swift
extension KingfisherManager {
	/**
	- Throws: `KingfisherError`
	*/
	@discardableResult
	func retrieveImage(
		with resource: Resource,
		options: KingfisherOptionsInfo? = nil,
		progressBlock: DownloadProgressBlock? = nil
	) async throws -> RetrieveImageResult {
		var downloadTask: DownloadTask?

		return try await withTaskCancellationHandler(
			handler: { [downloadTask] in
				downloadTask?.cancel()
			}
		) {
			// Needed because Kingfisher can in some cases incorrectly call the completion handler multiple times.
			var hasBeenCalled = false

			return try await withCheckedThrowingContinuation { continuation in
				downloadTask = retrieveImage(
					with: resource,
					options: options,
					progressBlock: progressBlock
				) { result in
					guard !hasBeenCalled else {
						return
					}

					continuation.resume(with: result)
					hasBeenCalled = true
				}
			}
		}
	}
}
```

Another example: https://gist.github.com/sindresorhus/b37c9ab01b79b754ca1690aaa266c410

**Comment by onevcat** [comment]
Hi, @sindresorhus 

Thanks for opening it. It is a great first step to wrap these block-based methods with a continuation and provides an async version of APIs. The migration is already on the plan list, but with several considerations. I will post some thoughts here for discussion:

- For general purpose methods like `KingfisherManager.retrieveImage`, it is proper to adopt to Swift concurrency, since the main target of these method is retrieving something and the completion block is important. Making it an async method would provides a much nicer interface.
- However, for methods like `UIImageView.kf.setImage` and other UIKit/AppKit view extension methods, it is **not a good idea to adopt to async method**. These methods have a target of providing an easy way to set the image. The completion block is only a "side-effect" to give the users a chance to access the result. It is not the main purpose. Adopting these methods to async function will make the use of Kingfisher ugly (thinking about adding a `Task` context every time you want to set an image to a view!). So I guess these methods won't have an async version to keep things clean.
- For iOS 13 and 14, there is no Foundation support for Swift concurrency. Or in another word, there is no async version of `URLSession` or file related APIs. So on these back-port versions, we are not able to get benefit from the cooperative dispatch queue. The only way is wrapping the network related things in a continuation. Also, that means even the Swift concurrency is back-ported to iOS 13, to get the most performance in iOS 15 or later, we still need to implement two sets of network or image processing under the hood.

As indicated by Apple, the full migration is expected to happen in several years, until we can achieve to a fully concurrency-compatible version eventually (at least when we can start supporting from iOS 15). As the first step, we can gradually provide some APIs an async version, so the users can get the good points of actors and structured concurrency at least. I will try to see what I can do for it.

**Comment by sindresorhus** [comment]
> However, for methods like UIImageView.kf.setImage and other UIKit/AppKit view extension methods, it is not a good idea to adopt to async method. These methods have a target of providing an easy way to set the image.

You could still provide async overloads though. The overload resolution works so that it will prioritize the async overload when in an async context and the sync overload when in a sync context.

https://github.com/apple/swift-evolution/blob/main/proposals/0296-async-await.md#overloading-and-overload-resolution

**Comment by onevcat** [comment]
Yes. It is no harm to have ones. I will try to start some basic work on it soon.

**Comment by flowbe** [comment]
Hello,
Xcode 13.2 and iOS 15.2 are officially released. Any updates on the advancement of this feature? It would be really cool

**Comment by alexanderkhitev** [comment]
Hello,
Any news/updates? Thanks.

**Comment by NikKovIos** [comment]
Any updates?

**Comment by gionoa** [comment]
👀

**Comment by NikKovIos** [comment]
Write a comment in order to prioritize this feature 👍

**Comment by onevcat** [comment]
I do not have much time to devote into Kingfisher recently due to my daily work and life. So this is a bit slow. However, the concurrency support is already under development in [this branch](https://github.com/onevcat/Kingfisher/tree/concurrency). 

But I need more time to make it actually useful and stable.

**Comment by zkhCreator** [comment]
[the branch](https://github.com/onevcat/Kingfisher/issues/1864#:~:text=under%20development%20in-,this%20branch,-.) is not existed yet

**Comment by onevcat** [comment]
For now, you can try with the `v8` branch, which is a candidate of the next major version of Kingfisher, and it contains necessary async methods supports.

I am doing the final checks with this new version and hopefully I can tag it very soon.

**Comment by VladimirHorky** [comment]
Any rough estimate when we can expect it? 
I tried to wrap `KingfisherManager.shared.retrieveImage()` into `withCheckedThrowingContinuation`, but it crashes with `SWIFT TASK CONTINUATION MISUSE: tried to resume its continuation more than once`

```
func thumbnails(url: URL) async throws -> UIImage {
        try await withCheckedThrowingContinuation { continuation in
            kingfisherDownloadTask?.cancel()
            kingfisherDownloadTask = KingfisherManager.shared.retrieveImage(with: url) { result in
                switch result {
                case .success(let retrieveImageResult):
                    continuation.resume(returning: retrieveImageResult.image)
                case .failure(let error):
                    continuation.resume(throwing: error)
                }
            }
        }
    }
 ```

**Comment by khoogheem** [comment]
@onevcat looks like `DownloadProgressBlock` should be sendable in the v8


Passing argument of non-sendable type 'DownloadProgressBlock?' (aka 'Optional<(Int64, Int64) -> ()>') outside of main actor-isolated context may introduce data races
